query/src/api/rpc/flight_client_stream.rs

```diff
@@ -52,11 +52,17 @@ impl FlightDataStream {
                         )
                     }
 
-                    let arrow_schema = Arc::new(schema.to_arrow());
-                    Ok(
-                        deserialize_batch(&flight_data, arrow_schema, true, &Default::default())
-                            .map(create_data_block)?,
+                    let (arrow_schema, ipc_schema) =
+                        deserialize_schemas(flight_data.data_body.as_slice()).unwrap();
+                    let arrow_schema = Arc::new(arrow_schema);
```

Comment:
Ahh, here.

An arrow stream is expected to contain the following:

`[schema][batch][batch]...[batch][batch]`

where batches are expected to be deserializable by `schema`.

I.e. we need to read the schema from the stream once (first message), and then read the multiple batches from it. I think that this is currently reading the stream as `[schema][batch][schema][batch]...[schema][batch][schema][batch]`

So, 

* read first message and store `schema` and `ipc_fields` (part of the [StreamMetadata](https://github.com/jorgecarleitao/arrow2/blob/main/src/io/ipc/read/stream.rs#L19))
* read next messages (batch messages) using `schema` and `ipc_fields` to deserialize them

Reply:
I fixed it, thank you very much for your help